### PR TITLE
refs #1728 fixed a bug where invalid characters in the port specifica…

### PR DESCRIPTION
**Pull request by davidnich**
…tion in a URL were ignored

### CODE REVIEW

File: lib/QoreURL.cpp [hunk 1]
```diff
@@ -65,52 +78,75 @@ struct qore_url_private {
...
+         //printd(5, "QoreURL::nbuf: '%s' size: %d\n", nbuf.c_str(), nbuf.size());
       }
       else
-	 nbuf = strdup(pos);
+	 nbuf.assign(pos);
 
       // see if there's a username
       // note that nbuf here has already had the path removed so we can safely do a reverse search for the '@' sign
-      if ((p = strrchr(nbuf, '@'))) {
+      if ((p = strrchr((char*)nbuf.c_str(), '@'))) {
```
Comment (pavelkveton) [inline lib/QoreURL.cpp]: why cast to (char*) here? `strrchr` should take `const char*` acc to the man page, so `nbuf.c_str()` should be enough...

File: lib/QoreURL.cpp [hunk 2]
```diff
@@ -65,52 +78,75 @@ struct qore_url_private {
...
 	 pos = p + 1;
+         // we terminate the string internally here, so we can no longer use string::size()
 	 *p = '\0';
 	 // see if there's a password
-	 if ((p = strchr(nbuf, ':'))) {
-	    printd(5, "QoreURL::parse_intern password=%s\n", p + 1);
+	 if ((p = strchr((char*)nbuf.c_str(), ':'))) {
```
Comment (pavelkveton) [inline lib/QoreURL.cpp]: same as above - why casting to `char*`?

File: lib/QoreURL.cpp [hunk 3]
```diff
@@ -65,52 +78,75 @@ struct qore_url_private {
...
+      // use std::string for a temporary buffer for exception-safety
+      std::string nbuf;
 
       // find end of hostname
-      if ((p = (char* )strchr(pos, '/'))) {
+      if ((p = (char*)strchr(pos, '/'))) {
 	 // get pathname if not at EOS
 	 path = new QoreStringNode(p);
-	 printd(5, "QoreURL::parse_intern path=%s\n", path->getBuffer());
+	 printd(5, "QoreURL::parse_intern path: '%s'\n", path->getBuffer());
```
Comment (pavelkveton) [inline lib/QoreURL.cpp]: here is a tab

File: lib/QoreURL.cpp [hunk 4]
```diff
@@ -65,52 +78,75 @@ struct qore_url_private {
...
 	 // get copy of hostname string for localized searching and invasive parsing
-	 nbuf = (char* )malloc(sizeof(char) * (p - pos + 1));
-	 strncpy(nbuf, pos, p - pos);
-	 nbuf[p - pos] = '\0';
+         nbuf.assign(pos, p - pos);
```
Comment (pavelkveton) [inline lib/QoreURL.cpp]: but here are spaces

**Review by pavelkveton — DISMISSED** [review]

File: lib/QoreURL.cpp (hunk 1 above)
Comment (davidnich) [reply lib/QoreURL.cpp]: without the cast, compilation fails with g++ as I wrote in the commit message

File: lib/QoreURL.cpp (hunk 2 above)
Comment (davidnich) [reply lib/QoreURL.cpp]: without the cast, compilation fails with g++ as I wrote in the commit message

File: lib/QoreURL.cpp (hunk 3 above)
Comment (davidnich) [reply lib/QoreURL.cpp]: fixed

**Comment by pavelkveton** [comment]
I see there is an error in man page on `strrchr` - the input is not `const char*`, actually. Anyway, I bumped to another problem: the function header in question is
```
DLLLOCAL void parse_intern(const char* buf, bool keep_brackets, ExceptionSink* xsink) {
```
but we soon kill the `const` in `buf` by using
```
char* p = strstr(buf, "://");
```
and later actively changing `p`. This is no go for me, even if the function is already released. I'll attempt to rewrite the function a bit.

**Pull request merged**